SocketRocket/SRWebSocket.m

```diff
@@ -563,14 +583,27 @@ - (void)_failWithError:(NSError *)error;
     });
 }
 
-- (void)_writeData:(NSData *)data;
+- (void)_writeData:(NSData *)data
 {
-    [self assertOnWorkQueue];
+    [self _writeData:data complete:NULL];
+}
 
+- (void)_writeData:(NSData *)data complete:(SRSendCompleteBlock)complete {
+    [self assertOnWorkQueue];
+    
     if (_closeWhenFinishedWriting) {
+        if (complete) {
+            complete(SRErrorWithCodeDescription(2134, @"socket is closed"));
+        }
         return;
     }
-
+    
+    SRDataCallback *record = [[SRDataCallback alloc] init];
+    record.completeBlock = complete;
```

Comment:
This `completion` can be `nil` and we are going to waste precious (on iOS at least) CPU cycles for running all this math later.
What do you think if we don't add the `record` to the `_sendCallbacks` at all if `completion` is `nil`?
This is going to ease the work later as well.